xmrstak/backend/cpu/minethd.cpp

```diff
@@ -285,23 +289,43 @@ void minethd::consume_work()
 	globalStates::inst().inst().iConsumeCnt++;
 }
 
-minethd::cn_hash_fun minethd::func_selector(bool bHaveAes, bool bNoPrefetch)
+minethd::cn_hash_fun minethd::func_selector(bool bHaveAes, bool bNoPrefetch, bool useXMR)
 {
 	// We have two independent flag bits in the functions
 	// therefore we will build a binary digit and select the
 	// function as a two digit binary
-	// Digit order SOFT_AES, NO_PREFETCH
-
-	static const cn_hash_fun func_table[4] = {
-		cryptonight_hash<0x80000, MEMORY, false, false>,
-		cryptonight_hash<0x80000, MEMORY, false, true>,
-		cryptonight_hash<0x80000, MEMORY, true, false>,
-		cryptonight_hash<0x80000, MEMORY, true, true>
+	// Digit order SOFT_AES, NO_PREFETCH, MINER_ALGO
+
+	static const cn_hash_fun func_table[] = {
+#ifndef CONF_NO_XMR
+		cryptonight_hash<XMR_MASK, XMR_ITER, XMR_MEMORY, false, false>,
+		cryptonight_hash<XMR_MASK, XMR_ITER, XMR_MEMORY, false, true>,
+		cryptonight_hash<XMR_MASK, XMR_ITER, XMR_MEMORY, true, false>,
+		cryptonight_hash<XMR_MASK, XMR_ITER, XMR_MEMORY, true, true>
+#endif
+#if (!defined(CONF_NO_AEON)) && (!defined(CONF_NO_XMR))
+		,
+#endif
+#ifndef CONF_NO_AEON
+		cryptonight_hash<AEON_MASK, AEON_ITER, AEON_MEMORY, false, false>,
+		cryptonight_hash<AEON_MASK, AEON_ITER, AEON_MEMORY, false, true>,
+		cryptonight_hash<AEON_MASK, AEON_ITER, AEON_MEMORY, true, false>,
+		cryptonight_hash<AEON_MASK, AEON_ITER, AEON_MEMORY, true, true>
+#endif
 	};
 
-	std::bitset<2> digit;
+	std::bitset<3> digit;
```

Comment:
You misundestood the code above. Please refer to the comment, we are essentially counting in binary with true / false flags.

So:
```
false false   =  0
false true    =  1
true  false   =  2
true  true    = 3
````

For bitset of 3 you need to have 8 entries in the table, and count up to 8.

Reply:
I will add some comments to this code part.

Reply:
Ok I see it now, my bad, this part is ok as-is

Reply:
There are 8 value in the table if XMR-STAK_CURRENCY=all. Else there are 4 values in the table, and bit 3 is always set to 0.